### ISSUE TITLE
libgvirtus-plugin-cufft.so: undefined symbol

### ISSUE THREAD

**Issue by shrebhan** [cjg/GVirtuS#8]
Hello

When I try to run the backend executable I get following error - 

````
[root@gpu-6 ~]# GVirtuS/bin/gvirtus-backend GVirtuS/etc/properties.json
INFO - 🛈  - GVirtuS backend version
INFO - 🛈  - Configuration: GVirtuS/etc/properties.json
INFO - 🛈  - Up and running
ERROR - Error loading: /lib/libgvirtus-plugin-cufft.so: undefined symbol: _Z26handleSetCompatibilityModeP12CufftHandlerSt10shared_ptrIN7gvirtus13communicators6BufferEE
INFO - 🛈  - [Process 6000] Shutdown
[root@gpu-6 ~]#
````

cuda version is 11.7
os - centos 8
cudnn version 8

@cjg @raffmont @vincenzosantopietro  @CarmineFerraro @anmentone

**Comment by Lilyse** [comment]
I run into same error.
you can use this command:
`ldd -r /lib/libgvirtus-plugin-cufft.so` 

I guess that there is some funtion not defined in this file

**Comment by shrebhan** [comment]
> I run into same error. you can use this command: `ldd -r /lib/libgvirtus-plugin-cufft.so`
> 
> I guess that there is some funtion not defined in this file

Yes there's quite a few undefined functions for me, probably a version mismatch

**Comment by shrebhan** [comment]
@Lilyse i was wondring if you had this issue too on the frontend ? - https://github.com/cjg/GVirtuS/issues/9

**Comment by Lilyse** [comment]
> @Lilyse i was wondring if you had this issue too on the frontend ? - #9

i maybe do same thing as you do.
i am trying to connect tensorflow with Gvirtus.

now get same error libcuda.so.1
`>>> tf.test.is_gpu_available()
2022-09-15 03:11:17.198275: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz
2022-09-15 03:11:17.205015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x6f09210 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2022-09-15 03:11:17.205059: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2022-09-15 03:11:17.207580: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2022-09-15 03:11:17.210803: E tensorflow/stream_executor/cuda/cuda_driver.cc:282] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2022-09-15 03:11:17.210837: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist
False
>>> exit()`

**Comment by Lilyse** [comment]
> > I run into same error. you can use this command: `ldd -r /lib/libgvirtus-plugin-cufft.so`
> > I guess that there is some funtion not defined in this file
> 
> Yes there's quite a few undefined functions for me, probably a version mismatch

i add some empty function like this
```
extern "C" __host__ cudaError_t CUDARTAPI cudaStreamEndCapture(cudaStream_t stream, cudaGraph_t *pGraph)
{
  cerr << "*** Error:cudaStreamEndCapture() not yet implemented!" << endl;
  return cudaErrorUnknown;
}
```
you need to compare the Gvirtus to the real cuda.
this may help

**Comment by shrebhan** [comment]
> > @Lilyse i was wondring if you had this issue too on the frontend ? - #9
> 
> i maybe do same thing as you do. i am trying to connect tensorflow with Gvirtus.
> 
> now get same error libcuda.so.1 `>>> tf.test.is_gpu_available() 2022-09-15 03:11:17.198275: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz 2022-09-15 03:11:17.205015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x6f09210 initialized for platform Host (this does not guarantee that XLA will be used). Devices: 2022-09-15 03:11:17.205059: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): Host, Default Version 2022-09-15 03:11:17.207580: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1 2022-09-15 03:11:17.210803: E tensorflow/stream_executor/cuda/cuda_driver.cc:282] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected 2022-09-15 03:11:17.210837: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist False
> 
> > > > exit()`


yes I also want to do the same thing, use tensorflow with gvirtus but I am trying with simple cuda binary first.
once we import the correct path on the front-end it is supposed to send it to backend to execute right ? That is not happening here ... could be something wrong in the properties.json ??

**Comment by Lilyse** [comment]
did you try this command
`nvcc matrixMul.cu --cudart=shared -lcuda -lcudart`
or 
`export EXTRA_NVCCFLAGS="--cudart=shared -lcuda -lcudart"`

notice:env without cuda

**Comment by Lilyse** [comment]
> > > @Lilyse i was wondring if you had this issue too on the frontend ? - #9
> > 
> > 
> > i maybe do same thing as you do. i am trying to connect tensorflow with Gvirtus.
> > now get same error libcuda.so.1 `>>> tf.test.is_gpu_available() 2022-09-15 03:11:17.198275: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz 2022-09-15 03:11:17.205015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x6f09210 initialized for platform Host (this does not guarantee that XLA will be used). Devices: 2022-09-15 03:11:17.205059: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): Host, Default Version 2022-09-15 03:11:17.207580: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1 2022-09-15 03:11:17.210803: E tensorflow/stream_executor/cuda/cuda_driver.cc:282] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected 2022-09-15 03:11:17.210837: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist False
> > > > > exit()`
> 
> yes I also want to do the same thing, use tensorflow with gvirtus but I am trying with simple cuda binary first. once we import the correct path on the front-end it is supposed to send it to backend to execute right ? That is not happening here ... could be something wrong in the properties.json ??

did you try this command
nvcc matrixMul.cu --cudart=shared -lcuda -lcudart
or
export EXTRA_NVCCFLAGS="--cudart=shared -lcuda -lcudart"

notice:env without cuda

**Comment by Lilyse** [comment]
> > > @Lilyse i was wondring if you had this issue too on the frontend ? - #9
> > 
> > 
> > i maybe do same thing as you do. i am trying to connect tensorflow with Gvirtus.
> > now get same error libcuda.so.1 `>>> tf.test.is_gpu_available() 2022-09-15 03:11:17.198275: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz 2022-09-15 03:11:17.205015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x6f09210 initialized for platform Host (this does not guarantee that XLA will be used). Devices: 2022-09-15 03:11:17.205059: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): Host, Default Version 2022-09-15 03:11:17.207580: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1 2022-09-15 03:11:17.210803: E tensorflow/stream_executor/cuda/cuda_driver.cc:282] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected 2022-09-15 03:11:17.210837: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist False
> > > > > exit()`
> 
> yes I also want to do the same thing, use tensorflow with gvirtus but I am trying with simple cuda binary first. once we import the correct path on the front-end it is supposed to send it to backend to execute right ? That is not happening here ... could be something wrong in the properties.json ??

did you try this command
nvcc matrixMul.cu --cudart=shared -lcuda -lcudart
or
export EXTRA_NVCCFLAGS="--cudart=shared -lcuda -lcudart"

notice:env without cuda

**Comment by Lilyse** [comment]
> > > @Lilyse i was wondring if you had this issue too on the frontend ? - #9
> > 
> > 
> > i maybe do same thing as you do. i am trying to connect tensorflow with Gvirtus.
> > now get same error libcuda.so.1 `>>> tf.test.is_gpu_available() 2022-09-15 03:11:17.198275: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz 2022-09-15 03:11:17.205015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x6f09210 initialized for platform Host (this does not guarantee that XLA will be used). Devices: 2022-09-15 03:11:17.205059: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): Host, Default Version 2022-09-15 03:11:17.207580: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1 2022-09-15 03:11:17.210803: E tensorflow/stream_executor/cuda/cuda_driver.cc:282] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected 2022-09-15 03:11:17.210837: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist False
> > > > > exit()`
> 
> yes I also want to do the same thing, use tensorflow with gvirtus but I am trying with simple cuda binary first. once we import the correct path on the front-end it is supposed to send it to backend to execute right ? That is not happening here ... could be something wrong in the properties.json ??

did you try this command
nvcc matrixMul.cu --cudart=shared -lcuda -lcudart
or
export EXTRA_NVCCFLAGS="--cudart=shared -lcuda -lcudart"

**Comment by shrebhan** [comment]
> > > > @Lilyse i was wondring if you had this issue too on the frontend ? - #9
> > > 
> > > 
> > > i maybe do same thing as you do. i am trying to connect tensorflow with Gvirtus.
> > > now get same error libcuda.so.1 `>>> tf.test.is_gpu_available() 2022-09-15 03:11:17.198275: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz 2022-09-15 03:11:17.205015: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x6f09210 initialized for platform Host (this does not guarantee that XLA will be used). Devices: 2022-09-15 03:11:17.205059: I tensorflow/compiler/xla/service/service.cc:176] StreamExecutor device (0): Host, Default Version 2022-09-15 03:11:17.207580: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1 2022-09-15 03:11:17.210803: E tensorflow/stream_executor/cuda/cuda_driver.cc:282] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected 2022-09-15 03:11:17.210837: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist False
> > > > > > exit()`
> > 
> > 
> > yes I also want to do the same thing, use tensorflow with gvirtus but I am trying with simple cuda binary first. once we import the correct path on the front-end it is supposed to send it to backend to execute right ? That is not happening here ... could be something wrong in the properties.json ??
> 
> did you try this command nvcc matrixMul.cu --cudart=shared -lcuda -lcudart or export EXTRA_NVCCFLAGS="--cudart=shared -lcuda -lcudart"

Yes I'm compiling the binary with all the flags, still same error - did it work for you?

**Comment by raffmont** [comment]
Maybe is better using this: https://github.com/gvirtus/GVirtuS

This is the currently maintained version of GVirtuS.

Please, try if you have the same issue.
I'll try to help you to fix it.
Best.
Raffaele

**Comment by shrebhan** [comment]
> Maybe is better using this: https://github.com/gvirtus/GVirtuS
> 
> This is the currently maintained version of GVirtuS.
> 
> Please, try if you have the same issue. I'll try to help you to fix it. Best. Raffaele

the repo that you linked seems to have the same commit history as this one, but I'll still try it out
